File: backend/app/api/v1/routers/safety.py
"""安全・品質管理APIルーター"""

import math
import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.rbac import UserRole, require_roles
from app.db.base import get_db
from app.models.user import User
from app.schemas.common import ApiResponse, PaginatedResponse, PaginationMeta
from app.schemas.safety import (
    QualityInspectionCreate,
    QualityInspectionResponse,
    SafetyCheckCreate,
    SafetyCheckResponse,
)
from app.services.safety_service import SafetyService

router = APIRouter(tags=["安全・品質管理"])


@router.post(
    "/projects/{project_id}/safety-checks",
    response_model=ApiResponse[SafetyCheckResponse],
    status_code=201,
)
async def create_safety_check(
    project_id: uuid.UUID,
    payload: SafetyCheckCreate,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[
        User,
        Depends(
            require_roles(
                UserRole.ADMIN, UserRole.PROJECT_MANAGER, UserRole.SITE_SUPERVISOR
            )
        ),
    ],
):
    """安全確認チェックリストを記録します。

    - 1 日 1 案件につき複数回記録可能
    - `items`: チェック項目ごとに `category`, `checked` (bool), `note` を含む配列
    - 権限: SITE_SUPERVISOR 以上
    """
    svc = SafetyService(db)
    data = await svc.create_safety_check(
        project_id, payload, created_by=current_user.id
    )
    return ApiResponse(data=data)


@router.get(
    "/projects/{project_id}/safety-checks",
    response_model=PaginatedResponse[SafetyCheckResponse],
)
async def list_safety_checks(
    project_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[
        User,
        Depends(
            require_roles(
                UserRole.ADMIN,
                UserRole.PROJECT_MANAGER,
                UserRole.SITE_SUPERVISOR,
                UserRole.VIEWER,
            )
        ),
    ],
    page: int = Query(default=1, ge=1),
    per_page: int = Query(default=20, ge=1, le=100),
):
    svc = SafetyService(db)
    items, total = await svc.list_safety_checks(project_id, page, per_page)
    return PaginatedResponse(
        data=items,
        meta=PaginationMeta(
            total=total,
            page=page,
            per_page=per_page,
            pages=math.ceil(total / per_page) if total > 0 else 0,
        ),
    )


@router.post(
    "/projects/{project_id}/quality-inspections",
    response_model=ApiResponse[QualityInspectionResponse],
    status_code=201,
)
async def create_quality_inspection(
    project_id: uuid.UUID,
    payload: QualityInspectionCreate,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[
        User,
        Depends(
            require_roles(
                UserRole.ADMIN, UserRole.PROJECT_MANAGER, UserRole.SITE_SUPERVISOR
            )
        ),
    ],
):
    svc = SafetyService(db)
    data = await svc.create_quality_inspection(
        project_id, payload, created_by=current_user.id
    )
    return ApiResponse(data=data)


@router.get(
    "/projects/{project_id}/quality-inspections",
    response_model=PaginatedResponse[QualityInspectionResponse],
)
async def list_quality_inspections(
    project_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[
        User,
        Depends(
            require_roles(
                UserRole.ADMIN,
                UserRole.PROJECT_MANAGER,
                UserRole.SITE_SUPERVISOR,
                UserRole.VIEWER,
            )
        ),
    ],
    page: int = Query(default=1, ge=1),
    per_page: int = Query(default=20, ge=1, le=100),
):
    svc = SafetyService(db)
    items, total = await svc.list_quality_inspections(project_id, page, per_page)
    return PaginatedResponse(
        data=items,
        meta=PaginationMeta(
            total=total,
            page=page,
            per_page=per_page,
            pages=math.ceil(total / per_page) if total > 0 else 0,
        ),
    )


@router.delete(
    "/projects/{project_id}/safety-checks/{check_id}",
    status_code=204,
)
async def delete_safety_check(
    project_id: uuid.UUID,
    check_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[
        User,
        Depends(
            require_roles(
                UserRole.ADMIN, UserRole.PROJECT_MANAGER, UserRole.SITE_SUPERVISOR
            )
        ),
    ],
):
    svc = SafetyService(db)
    await svc.delete_safety_check(project_id, check_id)


@router.delete(
    "/projects/{project_id}/quality-inspections/{inspection_id}",
    status_code=204,
)
async def delete_quality_inspection(
    project_id: uuid.UUID,
    inspection_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[
        User,
        Depends(
            require_roles(
                UserRole.ADMIN, UserRole.PROJECT_MANAGER, UserRole.SITE_SUPERVISOR
            )
        ),
    ],
):
    svc = SafetyService(db)
    await svc.delete_quality_inspection(project_id, inspection_id)

File: frontend/src/api/safety.ts
import api from "./client";
import type { PaginatedResponse } from "./projects";
import type {
  SafetyCheckResponse,
  SafetyCheckCreate,
  QualityInspectionResponse,
  QualityInspectionCreate,
} from "@/generated";

// Re-export with backward-compatible aliases
export type SafetyCheck = SafetyCheckResponse;
export type QualityInspection = QualityInspectionResponse;
export type { SafetyCheckCreate, QualityInspectionCreate };

export const safetyApi = {
  createSafetyCheck: async (projectId: string, data: SafetyCheckCreate) => {
    const res = await api.post<{ data: SafetyCheck }>(
      `/projects/${projectId}/safety-checks`,
      data
    );
    return res.data.data;
  },

  listSafetyChecks: async (projectId: string, page = 1, perPage = 20) => {
    const res = await api.get<PaginatedResponse<SafetyCheck>>(
      `/projects/${projectId}/safety-checks`,
      { params: { page, per_page: perPage } }
    );
    return res.data;
  },

  createQualityInspection: async (
    projectId: string,
    data: QualityInspectionCreate
  ) => {
    const res = await api.post<{ data: QualityInspection }>(
      `/projects/${projectId}/quality-inspections`,
      data
    );
    return res.data.data;
  },

  listQualityInspections: async (
    projectId: string,
    page = 1,
    perPage = 20
  ) => {
    const res = await api.get<PaginatedResponse<QualityInspection>>(
      `/projects/${projectId}/quality-inspections`,
      { params: { page, per_page: perPage } }
    );
    return res.data;
  },

  deleteSafetyCheck: async (projectId: string, checkId: string) => {
    await api.delete(`/projects/${projectId}/safety-checks/${checkId}`);
  },

  deleteQualityInspection: async (projectId: string, inspectionId: string) => {
    await api.delete(`/projects/${projectId}/quality-inspections/${inspectionId}`);
  },
};
